### PR TITLE
Closes #1908: Added Workbench Access module.

### LINKED ISSUE #1908: Add Workbench Access for section editing access control
## Motivation
As a site owner, I would like to delegate specific sections of my site to be edited by people that are not me or a member of my trusted staff.

[README for workbench_access ](https://git.drupalcode.org/project/workbench_access/-/blob/2.x/README.md)

### Is your feature request related to a problem? Please describe.
Keeping in mind that user access control can be a challenge at such a fluid institution where people come and go and switch departments, or graduate, it would be nice to be able to do this sort of thing on a limited basis.

It might also be to our benefit to create a rubrik for managing users who have section editing access, and maybe a boilerplate workflow and access request form to hand off to clients that find this useful.

## Proposed Resolution
Add workbench_access module and provide sane defaults, plus mimimal documentation for common use-cases.

### Describe alternatives you've considered
[organic groups](https://www.drupal.org/project/og)
[group](https://www.drupal.org/project/group)
[content access](https://www.drupal.org/project/content_access)
[Taxonomy Access Control Lite](https://www.drupal.org/project/tac_lite)

### Roles and Permissions considerations
A clear and concise description of how each of the following roles would be impacted by this change:
- User administrator (Can grant users with the section editor role)
- Section editor (A new role created)
- Content administrator (Determines the sections that are under access control)

### PR DESCRIPTION
**Pull request by cpoff**
## Description
This is simply adding the WB Access module to QS - no dependencies. Devs can enable as needed.

## Related issues
Closes #1908
#1972 is a duplicate

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [x] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by joeparsons — DISMISSED** [review]

**Comment by danahertzberg** [comment]
Should consider bundling default permissions with this module

https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-1973.probo.build/admin/people/permissions/module/workbench_access

**Comment by danahertzberg** [comment]
Would like to test out the UX and discuss these permissions as a potential default
image

**Review by trackleft — APPROVED** [review]

File: config/install/user.role.az_content_editor.yml [hunk 1]
```diff
@@ -138,3 +140,4 @@ permissions:
   - 'view webform revisions'
   - 'view webform submissions any node'
   - 'view webform submissions own node'
+
```
Comment (trackleft) [inline config/install/user.role.az_content_editor.yml:143]:
```suggestion
```

File: config/install/user.role.az_content_editor.yml [hunk 2]
```diff
@@ -137,4 +139,4 @@ permissions:
   - 'view the administration theme'
   - 'view webform revisions'
   - 'view webform submissions any node'
-  - 'view webform submissions own node'
+  - 'view webform submissions own node'
```
Comment (trackleft) [inline config/install/user.role.az_content_editor.yml:142]:
```suggestion
  - 'view webform submissions own node'
  
```

File: config/install/user.role.az_content_editor.yml (hunk 1 above)
Comment (trackleft) [inline config/install/user.role.az_content_editor.yml:143]:
```suggestion

```

**Review by trackleft — APPROVED** [review]

**Comment by danahertzberg** [comment]
Code updates are looking good. User role testing on probo is still needed to ensure default permissions are what we want.

**Review by joeparsons — APPROVED** [review]

**Comment by danahertzberg** [comment]
Testing this module seems to result in unintuitive circumstances. For example, I add a scheme that restricts access by menu section. When logging in as a user with that restriction designated, I can modify _any_ page but only place it within my designated menu section.

Another example is that when I restrict by taxonomy, I allow a user to access the News content type of any taxonomy. The results are that I can add any type of content, but another user without the News taxonomy restriction cannot add News items because it is restricted.

All in all, I'm thinking our current permissions are conflicting with what Workbench Access provides. I'm wondering if (more) new roles like "News admin" or "Event admin" would work better with Workbench Access?

There are 2 main use cases I am aware of that I was hoping WBA module would solve. 
1. User admins want to assign another user to only be able to add/edit AZ News content -- nothing else
2. User admins want to assign another user to only be able to edit or add pages to a designated menu section -- nowhere else

**Comment by danahertzberg** [comment]
Content admin cannot be used with WBA because of its "Bypass access control" permission. Need to look more into how our current roles interact with WBA config.

https://git.drupalcode.org/project/workbench_access

**Review by ejsamboy — DISMISSED** [review]

**Comment by trackleft** [comment]
My recommendation is that we include creating a section editor role as part of our documentation

**Comment by trackleft** [comment]
- [ ] Add new role as part this PR called section editor

**Comment by camikazegreen** [comment]
Why is this permission on Content Editors instead of Section Editors:
- Allow all members of this role to be assigned to Workbench Access sections
Lets users be assigned to editorial sections.

This permission seems like it should be only on Content Administrators:
- Batch update section assignments for content
Perform batch section assignments from the content overview page.

We should also try to continue the same pattern that we've used for our other permissions where the roles are additive instead of duplicative. If this role has to be duplicative (for example, if the Content Editor role is too permissive for the Section Editor role in some ways) then we need to be really careful about how we document this role. But hopefully we can find a way to follow the same pattern.

**Comment by camikazegreen** [comment]
It looks like Dana's comment was from before we added a separate Section Editor role, so some things will likely need to be adjusted.

**Comment by camikazegreen** [comment]
The configuration for the User Admin will need to be updated to allow the adding of this new role.

**Comment by cpoff** [comment]
Confab with Cameron, Eric, and Carris came up with following solution:

Module_Permissions___Quickstart_Review

-Remove Section Editor role entirely
-Remove the 'Batch update ...' permission from Content Editor
-Assign the 'Bypass...' permission to Content Admin
- The **Deny!** box (via  `Configuration > Workbench > Settings`) will remain  unchecked/not checked when module is enabled. It will be added as a requirement for when a WB Access Schema is created.

**Comment by camikazegreen** [comment]
I have verified that:

- Content Editors are not restricted in any way prior to enabling Workbench Access.
- All expected permissions are added when the module is enabled.
- Content Editors are _still_ not restricted until Access Schemes are configured.
- Only User Admins are able to access Workbench Access configuration
- Upon enabling Workbench Access, configuring a scheme and checking the Deny Access box, User Admin and Content Admin are still able to edit all content.
- Upon enabling Workbench Access, configuring a scheme and checking the Deny Access box, Content Editors are restricted from editing any pages that they have not been explicitly given permission for.
- Menu based Access Schemes allow for page inheritance in the way that you would expect them to. eg. If you are granted access to a top level page, you will be able to edit all children and grandchildren.
- Content Editors are still able to create pages, but do not have access to add it to the menu.
- If not added to their section of the menu they are not able to edit their own page.

Notes:

- The Edit tabs are removed for all un-allowed pages, except for the home page. Access is still restricted if you attempt to edit that page.
- One thing that is a bit unintuitive is that Content Admins will typically appear on the list that you can restrict access to, but you are not actually able to restrict their access. This is due to them inheriting the Content Editor role, so there isn't an easy way around this, but we can solve this with documentation and training.

**Review by camikazegreen — APPROVED** [review]

**Review by trackleft — COMMENTED** [review]
Sounds like the ducks are in alignment

**Review by trackleft — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**